### PR TITLE
escape_to_c: Escape trigraphs

### PR DESCRIPTION
**Pull request by jcbrinfo**
Produce more standard-compliant code instead of relying on the default behaviour of gcc (or on its `-Wno-trigraphs` flag).

See also: https://github.com/nitlang/nit/pull/1949/commits/000ac8de3c64790b0fbb0e868e406a6a280f3fa9

Signed-off-by: Jean-Christophe Beaupré <jcbrinfo@users.noreply.github.com>

### CODE REVIEW

File: lib/core/text/abstract_text.nit [hunk 1]
```diff
@@ -617,6 +620,17 @@ abstract class Text
 				b.append("\\\'")
 			else if c == '\\' then
 				b.append("\\\\")
+			else if c == '?' then
+				# Escape if it is the last question mark of a ANSI C trigraph.
+				var j = i + 1
+				if j < length then
+					var next = chars[j]
+					# We ignore `??'` because it will be escaped as `??\'`.
+					if ['=', '/', '(', ')', '!', '<', '>', '-'].has(next) then
```
Comment (privat) [inline lib/core/text/abstract_text.nit]: the array will be created for each '?' character (and iterated)

Comment (jcbrinfo) [inline lib/core/text/abstract_text.nit]: Sorry. I was (stupidly) assuming an optimization here. I will replace it by a bunch of `or`s.

File: lib/core/text/abstract_text.nit [hunk 2]
```diff
@@ -590,10 +590,13 @@ abstract class Text
 		return res.to_s
 	end
 
-	# Escape " \ ' and non printable characters using the rules of literal C strings and characters
+	# Escape `"` `\` `'`, [trigraphs](https://en.wikipedia.org/wiki/Digraphs_and_trigraphs#C) and non printable characters using the rules of literal C strings and characters
```
Comment (xymus) [inline lib/core/text/abstract_text.nit]: Please remove the link, it makes the pure text version of the synopsis way too long (for the doc integrated to vim) and we can google it ourselves.

**Comment by jcbrinfo** [comment]
:recycle: Reroll

**Comment by jcbrinfo** [comment]
I just found a potential side effect: it breaks `escape_to_nit`!

**Comment by jcbrinfo** [comment]
I just realized that `"\?" == "?"` in Nit. So, I need to do some work on this PR.

**Comment by jcbrinfo** [comment]
:recycle: Reroll, now with `FlatText` also fixed.

PS: Is there any objection to have question marks escaped in `escape_to_nit` and `escape_to_dot`?

**Comment by privat** [comment]
maybe extract from the `escape_to_c` a common base that could be used by the other escape_* ?

**Comment by jcbrinfo** [comment]
And lose the optimizations done by @R4PaSs? Not sure it is a good idea considering that many languages do not support the octal notation anyway.

File: lib/core/text/abstract_text.nit [hunk 3]
```diff
@@ -602,7 +605,13 @@ abstract class Text
...
 	#
 	# The exceptions are the common `\t` and `\n`.
-	fun escape_to_c: String
+	fun escape_to_c: String do return escape_to_c_text_impl
+
+	# The original definition of `escape_to_c`
+	#
+	# This method is separated from `escape_to_c` mainly for testing purposes.
+	# Normally, you should not call this method explicitly nor redefine it.
+	protected fun escape_to_c_text_impl: String
```
Comment (R4PaSs) [inline lib/core/text/abstract_text.nit]: Honestly I'm not a fan of having a new protected method just to add a test suite, would it be possible not to change the methods within text for testing and just have the contents of `text_abstract_text` be simple nitunits?

Comment (jcbrinfo) [inline lib/core/text/abstract_text.nit]: An alternative way might be to split `flat` in 2 modules: one for basic features and another for optimizations. In that case, the test must be a `test.sh` test in order to bypass nitunit’s imports.

Comment (R4PaSs) [inline lib/core/text/abstract_text.nit]: I really don't see why, if you want to test the method on receivers of each type, you can already do it by adding a test, instanciate each Text subclass, and check the result is really what you expected, there's no need to complicate an already complex hierarchy by either adding a proxy test method or splitting the modules.

File: lib/core/text/test_abstract_text.nit [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+import test_suite
+import flat
+
+redef class FlatText
+	# The examples in the documentation test the redefiniton in the (implicitly
+	# imported) `flat` module.
+	# Here, we force tests on the original definiton.
+
+	redef fun escape_to_c do return escape_to_c_text_impl
```
Comment (privat) [inline lib/core/text/test_abstract_text.nit]: You should not complexify an API for testing purpose.
I suggest that, in order to test the implementation of Text instead the ones of FlatText, to use a receiver that is a Text without being a FlatText.

Maybe something like:
~~~nit
intrude import core::ropes
var r = new Concat("??!??<", ??>??-")
assert r.escape_to_c == "?\\?!?\\?<?\\?>?\\?-"
~~~

**Comment by jcbrinfo** [comment]
:recycle: Reroll

**Comment by privat** [comment]
ping @R4PaSs

**Comment by R4PaSs** [comment]
I'm still not a big fan of the test suite, but the API and implementation look OK to me, :shipit:

**Comment by jcbrinfo** [comment]
Yeah, I had to workaround #2165.

**Comment by privat** [comment]
+1
#2165 is still an open issue :(

**Comment by xymus** [comment]
Please note that the description of this PR is misleading, the warning on nitc has already been fixed and it remains only in the bootstrap. This PR won't make it more disappear.

Still, I'm OK with the changes.

**Comment by jcbrinfo** [comment]
@xymus I will change the description to make it more accurate.

I’m thinking about that: should I remove the `-Wno-trigraph` flag since it is no longer needed?

**Comment by xymus** [comment]
@jcbrinfo It's not worth it, fixing the description was enough.

**Comment by jcbrinfo** [comment]
:recycle: reword

**Pull request merged**